### PR TITLE
cleanup helm chart

### PR DESCRIPTION
**Pull request by dungdm93**
This is next part of refactoring helm chart #16792 with the following changes:

1. remove `.Values.Capabilities` in favor of `--kube-version` in helm template,  as discussion in [#16752](https://github.com/cilium/cilium/pull/16752#discussion_r663973932)

2. Unify image using `cilium.image` template.
https://github.com/cilium/cilium/blob/c95f6fda172cd93bf2848b17dcc8bfdbe5ed63f4/install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml#L226
Currently, configuring image is quite long, and not consistent between components (some support digest, some are not). By using `cilium.image` template, it's make configuring image shorter, easier to read and consistent between components.

3. support `priorityClassName`.
Currently, some components (e.g agent) support define `priorityClassName` in values file, but not using in template :|, other components (e.g hubble-ui, hublle-relay) are not support `priorityClassName`. So, with this MR, `priorityClassName` is supported in all components.

4. move `hubble-metrics` service & `hubble`  ServiceMonitor to seperated file (in `hubble` folder)

5. remove unnecessary parentheses
Screenshot from 2021-07-15 20-30-29

6. make `name` as the first attribute of object
7. using `nindent` instead of `indent`
Screenshot from 2021-07-15 18-48-14

8. using default and ternary instead of if-else
Screenshot from 2021-07-15 18-47-15

9. using `with` instead of `if` when it's possible
Screenshot from 2021-07-23 15-40-35

Signed-off-by: Đặng Minh Dũng <dungdm93@live.com>

### CODE REVIEW

File: install/kubernetes/cilium/templates/_helpers.tpl
Comment (errordeveloper) [inline install/kubernetes/cilium/templates/_helpers.tpl]: As I mentioned in https://github.com/cilium/cilium/pull/16900#discussion_r672166934, this comment is confusing...

**Review by errordeveloper — COMMENTED** [review]
Thanks very much, overall this looks good to me!

**Review by gandro — CHANGES_REQUESTED** [review]
> remove redundant `namespace: {{ .Release.Namespace }}`

I mentioned it in the other PR, but this is unfortunately not redundant. This is needed for correct use of `helm template --namespace`, see https://github.com/helm/helm/issues/3553#issuecomment-521424132

We are relying on the fact that one can `kubectl create` the generated YAML without an explicit `-n` in the docs, so this will break existing workflows.

**Comment by dungdm93** [comment]
Thank @gandro for clearify. `namespace` is back

**Review by gandro — COMMENTED** [review]
From a first glance, this looks correct now, thank you! It will require some manual testing to ensure there are no significant differences in the output.

One thing we need to document (because it's a breaking change) is the removal of the `Capabilities` values. Please add an upgrade note here:
https://github.com/cilium/cilium/blob/master/Documentation/operations/upgrade.rst#111-upgrade-notes

It could be something like:

> - The `Capabilities` Helm value has been removed. When using `helm template` to generate the Kubernetes manifest for a specific Kubernetes version, please use the `--kube-version` flag (introduced in Helm 3.6.0) instead.

**Comment by dungdm93** [comment]
`v1.11 Upgrade Notes` added

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:397]: This `imagePullPolicy` seems to have gotten lost in the refactor

File: install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:149]: This is _not_ equivalent to `secretName: clustermesh-apiserver-server-cert`, as it will also mount in the CA private key (`tls.key`) in the pod, which I think we don't want.

File: install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:160]: Same as above, `secretName: clustermesh-apiserver-admin-cert` will mount the private key `tls.key`.

File: install/kubernetes/cilium/templates/_helpers.tpl
Comment (gandro) [inline install/kubernetes/cilium/templates/_helpers.tpl]: So with this template, you removed almost all uses of the `enableCriticalPriorityClass` value:
https://github.com/cilium/cilium/blob/master/install/kubernetes/cilium/values.yaml#L1550-L1555

Which I think is fine, as again in newer versions of Helm we can now use `--kube-version`. However, there is one use of of it left (which also relies on `.Values.Capabilities` that you removed):

https://github.com/cilium/cilium/blob/8c49b836cb101b897be95c06410125fb59cd18b5/install/kubernetes/cilium/templates/etcd-operator/cilium-etcd-operator-deployment.yaml#L98

Do you think it makes sense to replace that use as well, and remove `enableCriticalPriorityClass` entirely? And then also document it as being removed in the upgrade notes?

(I realize that the last use is in `cilium-etcd-operator` which you are probably planning to remove anyway, but I prefer if we were consistent even for code that is about to be removed)

**Review by gandro — COMMENTED** [review]
Starting to look good. A few minor things left that I noticed while comparing the `helm template` output

File: install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml
Comment (qmonnet) [inline install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml]: Nit: let's fix the typo, `in sync` or `synced` but not `in synced`?

File: install/kubernetes/cilium/templates/cilium-preflight/clusterrole.yaml
Comment (qmonnet) [inline install/kubernetes/cilium/templates/cilium-preflight/clusterrole.yaml]: Same here.

File: install/kubernetes/cilium/values.yaml
Comment (qmonnet) [inline install/kubernetes/cilium/values.yaml]: This ends up in the Documentation too, so maybe use something slightly more descriptive? Some similar fields use:
```
The priority class to use for ...
```

By the way, you must also build the documentation locally (`make -C Documentation`) and commit the changes from generated file Documentation/helm-values.rst. The CI will likely complain when it runs otherwise.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (qmonnet) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml]: Nit: I'd put the `do` higher to keep just the commands inside the block, but no strong preference.
```suggestion
          while true; do
            sleep 30;
          done
```

**Review by qmonnet — CHANGES_REQUESTED** [review]
Looks good overall, although my knowledge of the Helm chart is limited and I'm not able to provide an in-depth review of the changes.

Here are a few nitpicks below. In addition,
- Please make sure to update the Helm reference in the Documentation (see comment below),
- And please update your commit log message regarding the `remove redundant namespace: {{ .Release.Namespace }}`, given that, if I understand the discussion with Sebastian, this is in fact not the case.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (qmonnet) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:397]: Naive question, would it make sense to have the `imagePullPolicy` part of the new `cilium.image` template? I think they are always used together. Or is it impossible/bad practice to create two elements (`image` and `imagePullPolicy`) at once?

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (dungdm93) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:74]: > This imagePullPolicy seems to have gotten lost in the refactor
@gandro `imagePullPolicy` is here

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:397]: @gandro `imagePullPolicy` is not be forgotten. I just mention you in where it place.

File: install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml]: It's old word, but I update it as your suggestion :+1:

File: install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:149]: @gandro I think you have a little confuse here: CA private key is `ca.key`, not `tls.key`. `tls.key` is private key of certificate `tls.crt`

Comment (dungdm93) [reply install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:149]: Two approaches are the same, it contains 3 files `ca.crt`, `tls.crt`, `tls.key`. and no CA private key (`ca.key`).
```yaml
      - name: etcd-server-secrets
        projected:
          defaultMode: 0420
          sources:
          - secret:
              name: clustermesh-apiserver-ca-cert
              items:
              - key: ca.crt
                path: ca.crt
          - secret:
              name: clustermesh-apiserver-server-cert
```
mean the volume contains:
* `ca.crt` from `clustermesh-apiserver-ca-cert`, `ca.key` is not mount because it's not in `items` list
* mount all files from `clustermesh-apiserver-server-cert` except `ca.cert` because it's overridden by `ca.crt` from `clustermesh-apiserver-ca-cert`

And because `ca.crt` from `clustermesh-apiserver-ca-cert` and `clustermesh-apiserver-server-cert` are the same. So we can eliminate `clustermesh-apiserver-ca-cert` and simplify as bellow:
```yaml
      - name: etcd-server-secrets
        secret:
          secretName: clustermesh-apiserver-server-cert
          defaultMode: 0420
```

File: install/kubernetes/cilium/templates/_helpers.tpl
Comment (dungdm93) [reply install/kubernetes/cilium/templates/_helpers.tpl]: Oh, I miss `enableCriticalPriorityClass` check. I just update `cilium.priorityClass` template with this check and few changes in the usage. And `cilium-etcd-operator` is updated as well.

For `enableCriticalPriorityClass` removal, I totally agree, but let ask other people before we make the change.

File: install/kubernetes/cilium/values.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/values.yaml]: @qmonnet thanks. The docs is updated

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (gandro) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:74]: This is the `imagePullPolicy` for the `cilium-agent` container, not for the `mount-cgroup` where it was lost.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (gandro) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:397]: Maybe I'm misunderstanding something, but this here is the `mount-cgroup` initContainer. I don't see it in this container? The other one is for the `cilium-agent` container.

File: install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml
Comment (gandro) [reply install/kubernetes/cilium/templates/clustermesh-apiserver/deployment.yaml:149]: Ohh, you are 100% correct. Good catch - thank you!

File: install/kubernetes/cilium/templates/_helpers.tpl
Comment (gandro) [reply install/kubernetes/cilium/templates/_helpers.tpl]: Great change! Yes, agreed, let's keep `enableCriticalPriorityClass` in, now that it is used consistently. We can always remove it later.

**Review by qmonnet — APPROVED** [review]
Thanks for the changes. Pending the remaining concerns from Sebastian are addressed, this looks good to me :+1:

File: install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml
Comment (qmonnet) [reply install/kubernetes/cilium/templates/cilium-agent/clusterrole.yaml]: I see you did another commit just for that. I just meant updating the line in the same commit since you were already changing `-` to `/`, not to create a separate commit :) But OK keep it like this, and thanks for fixing the typo!

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:74]: Oh, thank @gandro. I miss this one.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml:397]: Thanks @gando, it's added.

**Review by gandro — APPROVED** [review]
Awesome work! Thank you!

**Comment by gandro** [comment]
Travis on ARM failed with a transient error, should not block this PR: https://app.travis-ci.com/github/cilium/cilium/jobs/526175793
```
# Pre-pull FROM docker images due to Buildkit sometimes failing to pull them.

grep "^FROM " Dockerfile | cut -d ' ' -f2 | xargs -n1 docker pull

Error response from daemon: Get https://registry-1.docker.io/v2/library/python/manifests/3.7.9-alpine3.11: Get https://auth.docker.io/token?scope=repository%3Alibrary%2Fpython%3Apull&service=registry.docker.io: EOF
```

**Comment by gandro** [comment]
test-me-please

**Review by christarazi — COMMENTED** [review]
LGTM. Could you squash `Update some typo` into the first commit? It fits the description of the first commit anyway.

Looks like you'll also need to rebase on master.

**Comment by dungdm93** [comment]
@christarazi rebased, and `Update some typo` also be squashed.

File: install/kubernetes/cilium/templates/cilium-operator/deployment.yaml
Comment (christarazi) [inline install/kubernetes/cilium/templates/cilium-operator/deployment.yaml]: This change is supposed to be part of the 1st commit but is part of the 3rd?

File: install/kubernetes/cilium/templates/cilium-preflight/daemonset.yaml
Comment (christarazi) [inline install/kubernetes/cilium/templates/cilium-preflight/daemonset.yaml]: This change is supposed to be part of the 1st commit but is part of the 3rd?

File: install/kubernetes/cilium/templates/etcd-operator/cilium-etcd-operator-deployment.yaml
Comment (christarazi) [inline install/kubernetes/cilium/templates/etcd-operator/cilium-etcd-operator-deployment.yaml:21]: This change is supposed to be part of the 1st commit but is part of the 3rd?

**Review by christarazi — COMMENTED** [review]
A few more git history related comments.

File: install/kubernetes/cilium/templates/cilium-operator/deployment.yaml
Comment (dungdm93) [reply install/kubernetes/cilium/templates/cilium-operator/deployment.yaml]: @christarazi  The 1st commit also contains `priorityClassName` changes, so I decided to squash 3rd commit into 1st commit.
I also update 1st commit message to match w/ this PR description.

**Review by aanm — APPROVED** [review]
Thank you for this amazing PR. Two comments:

1. Is there any linter that you used to catch all of these inconsistencies?
2. Just to confirm, all changes from https://github.com/cilium/cilium/pull/16933 seem to be included in this PR, correct?

**Comment by dungdm93** [comment]
@aanm
1. Nope, I check all of this with my eye :)
2. Yeah, all changes from #16933 are coverd in this PR

**Comment by aanm** [comment]
test-me-please

**Review by christarazi — APPROVED** [review]
:rocket:

**Comment by dungdm93** [comment]
@christarazi I already rebase master, but `needs-rebase` isn't auto remove.

**Comment by dungdm93** [comment]
test-me-please

**Comment by dungdm93** [comment]
@christarazi I rebased form master. But seem `needs-rebase` tags is not auto remove

**Review by qmonnet — APPROVED** [review]
Not sure why you need another review from me, but it still looks good.

**Comment by dungdm93** [comment]
@qmonnet sorry, I click to request review from @errordeveloper but accident click into your name :)

**Comment by qmonnet** [comment]
test-me-please

**Comment by qmonnet** [comment]
Ilya is currently out of office, given that his feedback has been addressed I think we can take [his earlier comment](https://github.com/cilium/cilium/pull/16896#pullrequestreview-709397177) as an approval. I'm unassigning him.

**Comment by qmonnet** [comment]
Considering Ilya's implicit approval (see above) and that all tests have passed, I'm labelling as `ready-to-merge`.

**Pull request merged**
